### PR TITLE
Hash identities

### PR DESCRIPTION
**Pull request by reginafcompton**
This PR adds a management command that hashes user identities (phone numbers). It hashes identities if:
1. they canceled or completed their session more than 5 minutes ago;
2. they abandoned their session more than 24 hours ago.

This PR works in cooperation with https://github.com/datamade/rapidsms/pull/3/files

It handles issue https://github.com/datamade/coordinated-entry-screening/issues/67

@fgregg - ready for your review after the holidays!

### CODE REVIEW

File: ces_admin/management/commands/hash_identities.py [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+from decisiontree.models import Session
+
+class Command(BaseCommand):
+    help = 'Hashes the user identities, i.e., phone numbers'
+
+    def handle(self, *args, **options):
+        five_minutes_ago = datetime.datetime.now() - datetime.timedelta(minutes=5)
+        one_day_ago = datetime.datetime.now() - datetime.timedelta(days=1)
+        # Find connections from sessions that do not have a state_id, i.e., the user completed the survey or canceled their session
+        # Add a five minute boundary, in the rare event that Rapidsms or Twilio
```
Comment (fgregg) [inline ces_admin/management/commands/hash_identities.py]: seems like part of this sentence has been cut off?

File: ces_admin/management/commands/hash_identities.py [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+class Command(BaseCommand):
+    help = 'Hashes the user identities, i.e., phone numbers'
+
+    def handle(self, *args, **options):
+        five_minutes_ago = datetime.datetime.now() - datetime.timedelta(minutes=5)
+        one_day_ago = datetime.datetime.now() - datetime.timedelta(days=1)
+        # Find connections from sessions that do not have a state_id, i.e., the user completed the survey or canceled their session
+        # Add a five minute boundary, in the rare event that Rapidsms or Twilio 
+        # Exclude sessions that have already been hased or are websessions. 
+        filter_for_closed = Q(session__state_id__isnull=True, session__last_modified__lte=five_minutes_ago)
```
Comment (fgregg) [inline ces_admin/management/commands/hash_identities.py]: this variable name does not seem to line up with the filters?

File: ces_admin/management/commands/hash_identities.py [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+        filter_for_abandoned = Q(session__last_modified__lte=one_day_ago)
+
+        connections_from_closed_sessions = Connection.objects \
```
Comment (fgregg) [inline ces_admin/management/commands/hash_identities.py]: these are not all "closed sessions" so I'd use a different variable name. expired_or_closed perhaps

File: ces_admin/management/commands/hash_identities.py [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+                                                     .filter(filter_for_closed | filter_for_abandoned) \
+                                                     .filter(identity__startswith='+1') \
+                                                     .distinct()
+
+        for connection in connections_from_closed_sessions:
+            identity = connection.identity
+            hashed_identity = hashlib.sha256(identity.encode()).hexdigest()
```
Comment (fgregg) [inline ces_admin/management/commands/hash_identities.py]: I'd recommend using django's in-built hashers https://docs.djangoproject.com/en/2.1/topics/auth/passwords/

File: ces_admin/management/commands/hash_identities.py (hunk 1 above)
Comment (reginafcompton) [reply ces_admin/management/commands/hash_identities.py]: Why, yes. I've improved the comment overall....

File: ces_admin/management/commands/hash_identities.py (hunk 3 above)
Comment (reginafcompton) [reply ces_admin/management/commands/hash_identities.py]: changed it to something more general, but more accurate: `sms_connections`

File: ces_admin/management/commands/hash_identities.py (hunk 4 above)
Comment (reginafcompton) [reply ces_admin/management/commands/hash_identities.py]: all right, I used `make_password` from Django [(which can be implemented independent of the User model)](https://docs.djangoproject.com/en/1.11/topics/auth/passwords/#module-django.contrib.auth.hashers). 

I think it makes sense - however, on the Rapidsms side of things, I have to iterate over all connections and run `check_password`, to find previously hashed identities:
https://github.com/datamade/rapidsms/pull/3/files#diff-632733e9d1d3abf4062a87fc96a1a840R101 Could this become inefficient if we have, say, thousands of connections in the database?

File: ces_admin/management/commands/hash_identities.py [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+        # Use the `startswith` filter to exclude sessions that 
+        # have already been hased or are websessions. 
+        sms_connections = Connection.objects \
+                                    .filter(filter_for_canceled_or_completed | filter_for_abandoned) \
+                                    .filter(identity__startswith='+1') \
+                                    .distinct()
+
+        for connection in sms_connections:
+            identity = connection.identity
+            hashed_identity = make_password(identity)
```
Comment (fgregg) [inline ces_admin/management/commands/hash_identities.py]: at the top of this function, I'd do something like

```
# comment explaining what you are doing
hash_phone_number = make_password
```

and then here have the code be `hash_phone_number(identity)`

and similar in the `check_password`

File: ces_admin/management/commands/hash_identities.py [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+        # our app still needs access to the unhashed phone number.   
+        five_minutes_ago = datetime.datetime.now() - datetime.timedelta(minutes=5)
+        one_day_ago = datetime.datetime.now() - datetime.timedelta(days=1)
+
+        filter_for_canceled_or_completed = Q(session__state_id__isnull=True, \
+                                             session__last_modified__lte=five_minutes_ago)
+        filter_for_abandoned = Q(session__last_modified__lte=one_day_ago)
+
+        # Use the `startswith` filter to exclude sessions that 
+        # have already been hased or are websessions.
```
Comment (fgregg) [inline ces_admin/management/commands/hash_identities.py]: s/hased/hashed/

**Pull request merged**